parlai/agents/transformer/module_extensions.py

```diff
@@ -0,0 +1,195 @@
+                "sum",
+            )
+        ) / noised_logits.size(0)
+
+    def forward_pass_with_r3f(self, model, *args, **kwargs):
+        """
+        'model' argument here is the `nn.Module` normally responsible for the full
+        forward pass.
+        """
+        if not self.use_r3f:
+            raise RuntimeError(
+                "Trying to use an R3F loss when this has not been allowed."
+            )
+        regular_output = model(*args, **kwargs)
+        with R3FNoiseEmbeddingContextManager(self, model) as model:
+            noised_output = model(*args, **kwargs)
+        symm_kl = self._calculate_symm_kl(noised_output[0], regular_output[0])
+        return regular_output, self.r3f_lambda * symm_kl
+
+
+class R3FNoisedEmbedding(nn.Module):
+    def __init__(self, noise_sampler):
+        super(R3FNoisedEmbedding, self).__init__()
+        self.noise_sampler = noise_sampler
+
+    def __call__(self, module, input, output):
+        noise = self.noise_sampler.sample(sample_shape=output.shape).to(output)
+        return output + noise
+
+
+class R3FNoiseEmbeddingContextManager(AbstractContextManager):
+    """
+    This class finds the first embedding(s) under the encoder/decoder, depending on what
+    is selected, and adds R3F noise to their output.
+
+    Note that this does not discrinate between the dictionary embedding lookup +
+    position embedding classes. However, seeing as how addition is commutative, this
+    should not be an issue.
+    """
+
+    def __init__(self, context, module):
+        self.encoder_hook = None
+        self.decoder_hook = None
+        self.context = context
+        for name, layer in module.named_modules():
+            if context.noise_encoder and name == "encoder":
+                self._find_embedding_and_set_hook(layer, "encoder")
+            if context.noise_decoder and name == "decoder":
+                self._find_embedding_and_set_hook(layer, "decoder")
+        self.module = module
+        self.context.had_first_pass = True
+
+    def _find_embedding_and_set_hook(self, layer, name):
+        try:
+            parent, embedding_name = self._find_embedding(layer)
+        except TypeError:
+            raise RuntimeError(
+                f"R3F noise: Could not find an embedding as a child of {name}. Does the model you are training use this type of layer? (Default R3F settings noises the encoder only, which some models - ex. GPT2 - do not use.)"
+            )
+        if not self.context.had_first_pass:
+            # need to do a deep copy because models oftentimes share embedding object for encoder + decoder; breaks noise encoder/decoder only experiments
+            setattr(
+                parent, embedding_name, copy.deepcopy(getattr(parent, embedding_name))
+            )
+        embedding = getattr(parent, embedding_name)
+        val = R3FNoisedEmbedding(self.context.noise_sampler)
+        setattr(self, name + "_hook", embedding.register_forward_hook(val))
```

Comment:
Do we need to store this as a variably-named attribute, rather than just accumulating into a list like `self._hooks` or similar?

Reply:
Yeah, I realized today that just using a map or something is probably way nicer than this `setattr` mess I was doing last night.